test/end2end_test.go

```diff
@@ -6362,3 +6367,165 @@ func (s) TestGlobalBinaryLoggingOptions(t *testing.T) {
+	return ctx
+}
+func (h *statsHandlerRecordEvents) HandleRPC(_ context.Context, s stats.RPCStats) {
+	h.mu.Lock()
+	h.s = append(h.s, s)
+	h.mu.Unlock()
+}
+func (*statsHandlerRecordEvents) TagConn(ctx context.Context, _ *stats.ConnTagInfo) context.Context {
+	return ctx
+}
+func (*statsHandlerRecordEvents) HandleConn(context.Context, stats.ConnStats) {}
+
+type blockingPicker struct {
+	pickDone func()
+}
+
+func (bp *blockingPicker) Pick(pi balancer.PickInfo) (balancer.PickResult, error) {
+	defer bp.pickDone()
+	return balancer.PickResult{}, balancer.ErrNoSubConnAvailable
+}
+
+const name = "blockingPickerBalancer"
+
+type blockingPickerBalancerBuilder struct{}
+
+func (blockingPickerBalancerBuilder) Build(cc balancer.ClientConn, bOpts balancer.BuildOptions) balancer.Balancer {
+	b := &blockingPickerBalancer{
+		ClientConn: cc,
+	}
+	// round_robin child to complete balancer tree with a usable leaf policy and
+	// have RPCs actually work.
+	builder := balancer.Get(roundrobin.Name)
+	if builder == nil {
+		// Shouldn't happen, defensive programming. Registered from import of
+		// roundrobin package.
+		return nil
+	}
+	rr := builder.Build(b, bOpts)
+	if rr == nil {
+		// Shouldn't happen, defensive programming.
+		return nil
+	}
+	b.Balancer = rr
+	return b
+}
+
+func (blockingPickerBalancerBuilder) ParseConfig(json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	return &bpbConfig{}, nil
+}
+
+func (blockingPickerBalancerBuilder) Name() string {
+	return name
+}
+
+type bpbConfig struct {
+	serviceconfig.LoadBalancingConfig
+}
+
+type blockingPickerBalancer struct {
+	stateMu    sync.Mutex
+	childState balancer.State
+
+	blockingPickerDone *grpcsync.Event
+	// embed a ClientConn to wrap only UpdateState() operation
+	balancer.ClientConn
+	// embed a Balancer to wrap only UpdateClientConnState() operation
+	balancer.Balancer
+}
+
+func (bpb *blockingPickerBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
+	bpb.blockingPickerDone = grpcsync.NewEvent()
+	err := bpb.Balancer.UpdateClientConnState(balancer.ClientConnState{
+		ResolverState: s.ResolverState,
+	})
+	bpb.ClientConn.UpdateState(balancer.State{
+		ConnectivityState: connectivity.Connecting,
+		Picker: &blockingPicker{
+			pickDone: func() {
+				bpb.blockingPickerDone.Fire()
+				bpb.stateMu.Lock()
+				cs := bpb.childState
+				bpb.stateMu.Unlock()
+				bpb.ClientConn.UpdateState(cs)
+			},
+		},
+	})
+	return err
+}
+
+func (bpb *blockingPickerBalancer) UpdateState(state balancer.State) {
+	bpb.stateMu.Lock()
+	bpb.childState = state
+	bpb.stateMu.Unlock()
+	if bpb.blockingPickerDone.HasFired() { // guard first one to get a picker sending ErrNoSubConnAvailable first
+		bpb.ClientConn.UpdateState(state) // after the first rr picker update, cc will trigger more, so actually forward these
+	}
+}
+
+// TestPickerBlockingStatsCall tests the emission of a stats handler call that
+// represents the picker had to block due to ErrNoSubConnAvailable being
```

Comment:
Terminology nit: the picker doesn't block.  I guess the `blockingPicker` blocks.  But more accurately, the RPC blocks (or in C++/Java: is queued) waiting for a new picker from the LB policy.

Reply:
Switched language accordingly :).